index.js

```diff
@@ -64,15 +90,61 @@ function makeFSObject(p) {
   return makeFSObjectCleanedUp(cleanUpPath(p));
 }
 
-function makeFSObjectCleanedUp(p) {
-  let stats = fs.lstatSync(p);
-  if (stats.isDirectory()) return new Directory(p, true);
-  if (stats.isFile()) return new File(p, true, stats);
-  // Return FSObject pointing to target of symbolic link. This is so you can use
-  // the returned FSObject to create a symlink without symlink indirection
-  // growing out of control.
-  if (!stats.isSymbolicLink())
-    throw new Error("File has unexpected type: " + p);
+function getDirectoryClass(p) {
+  if (GLOBAL_CACHE.has(p)) {
+    return GLOBAL_CACHE.get(p);
+  }
+
+  let directory = new Directory(p, true);
+
+  GLOBAL_CACHE.set(p, directory);
+
+  return directory;
+}
+
+function getFileClass(p, stats) {
+  if (GLOBAL_CACHE.has(p)) {
+    return GLOBAL_CACHE.get(p);
+  }
+
+  let directory = new File(p, true, stats);
+
+  GLOBAL_CACHE.set(p, directory);
+
+  return directory;
+}
+
+function makeFSObjectCleanedUp(p, dirent) {
+  if (dirent) {
+    if (dirent.isDirectory()) {
+      return getDirectoryClass(p);
+    }
+
+    if (dirent.isFile()) {
+      return getFileClass(p);
+    }
+
+    if (!dirent.isSymbolicLink()) {
+      throw new Error("File has unexpected type: " + p);
+    }
+  } else {
+    let stats = fs.lstatSync(p);
```

Comment:
any difference in perf in sync vs async?

Reply:
likely yes (for SSD drives, because there is  more file threads available), but this package is sync

Reply:
but for stat it's not clear - https://github.com/nodejs/node/issues/38006